source/globals.d.ts

```diff
@@ -76,10 +80,20 @@ declare module "delegate" {
 }
 
 // TODO: Add to dom-chef types
```

Comment:
Don't mind my scrap book of comments here. I'm still trying to figure out the best way to handle JSX with types to convert to regular HTML elements.

JSX gets returned as type `JSX.Element` which obviously does not map to `HTMLElement`, `SVGElement`, `HTMLAnchorElement` etc.

The worst case which I already mentioned initially in #1735, is to have a function which is essentially an identity function but will type things correctly, e.g.

```
function jsxToMarkup<T extends Element = HTMLElement>(jsx: JSX.Element): T {
	// Need to cast to unknown first as direct cast will not work.
	return (jsx as unknown) as T;
}
```

And then use it like so:

```
const anchor = jsxToMarkup<HTMLAnchorElement>(<a href={url}>{text}</a>);
```

I'm still going to see if I can come up with something better. If you guys have some suggestions, I'm all ears.

Reply:
The helper function is a no-go imho.

> Intrinsic elements are looked up on the special interface JSX.IntrinsicElements. By default, if this interface is not specified, then anything goes and intrinsic elements will not be type checked

What's the problem exactly? Can I see what error TS spits out? Can this be solved with `<a>1</a> as HTMLAnchorElement`?

Reply:
The problem isn't the intrinsic elements. As you can see in the PR, I've added the web components that are used in RGH there and they get recognized fine by the TypeScript. The problem is the `JSX.Element` interface. All snippets of JSX are of that type. I'm on mobile at the moment, but will explain later. But the idea is to do like you said, have it so they get typed to their corresponding HTML element type, E.g. `HTMLAnchorElement`

Reply:
We could just do

```
declare namespace JSX {
   ...
   type Element = any;
   ...
}
```

And then any `JSX.Element` is just `any` and it will be accepted anywhere, but this seems to loose to me. I'll dig a bit more tomorrow.

Reply:
A union type does not work from my initial tests. I don't think it's able to distinguish which type the JSX is in the union type of `JSX.Element`.

![image](https://user-images.githubusercontent.com/833231/53317108-65771c00-3898-11e9-80ad-574769f83f97.png)